dart/constraint/ContactSurface.hpp

```diff
@@ -0,0 +1,162 @@
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_CONSTRAINT_CONTACTSURFACE_HPP
+#define DART_CONSTRAINT_CONTACTSURFACE_HPP
+
+#include <Eigen/Core>
+#include "dart/collision/Contact.hpp"
+#include "dart/dynamics/ShapeNode.hpp"
+#include "dart/constraint/SmartPointer.hpp"
+
+namespace dart {
+namespace constraint {
+
+#define DART_RESTITUTION_COEFF_THRESHOLD 1e-3
+#define DART_FRICTION_COEFF_THRESHOLD 1e-3
+#define DART_BOUNCING_VELOCITY_THRESHOLD 1e-1
+#define DART_MAX_BOUNCING_VELOCITY 1e+2
+#define DART_CONTACT_CONSTRAINT_EPSILON_SQUARED 1e-12
+
+constexpr double DART_DEFAULT_FRICTION_COEFF = 1.0;
+constexpr double DART_DEFAULT_RESTITUTION_COEFF = 0.0;
+// slip compliance is combined through addition,
+// so set to half the global default value
+constexpr double DART_DEFAULT_SLIP_COMPLIANCE = 0.0;
+const Eigen::Vector3d DART_DEFAULT_FRICTION_DIR = Eigen::Vector3d::UnitZ();
+const Eigen::Vector3d DART_DEFAULT_CONTACT_SURFACE_MOTION_VELOCITY = Eigen::Vector3d::Zero();
+
+/// Computed parameters of the contact surface
+struct ContactSurfaceParams
+{
+  /// Primary Coefficient of Friction
+  double mFrictionCoeff {DART_DEFAULT_FRICTION_COEFF};
+
+  /// Secondary Coefficient of Friction
+  double mSecondaryFrictionCoeff {DART_DEFAULT_FRICTION_COEFF};
+
+  /// Coefficient of restitution
+  double mRestitutionCoeff {DART_DEFAULT_RESTITUTION_COEFF};
+
+  /// Primary Coefficient of Slip Compliance
+  double mSlipCompliance {DART_DEFAULT_SLIP_COMPLIANCE};
+
+  /// Secondary Coefficient of Slip Compliance
+  double mSecondarySlipCompliance {DART_DEFAULT_SLIP_COMPLIANCE};
+
+  /// First frictional direction (in world frame)
+  Eigen::Vector3d mFirstFrictionalDirection {DART_DEFAULT_FRICTION_DIR};
+
+  /// Velocity of the contact independent of friction
+  /// x = vel. in direction of contact normal
+  /// y = vel. in first friction direction
+  /// z = vel. in second friction direction
+  Eigen::Vector3d mContactSurfaceMotionVelocity {DART_DEFAULT_CONTACT_SURFACE_MOTION_VELOCITY};
+};
```

Comment:
Do you mean hiding the whole struct's contents in a PIMPL and adding public non-virtual getters/setters to the struct? I'm not sure what could be the performance impact of creating and destroying the unique pointers, as this structure is pretty short-lived and there can theoretically be thousands of them created every simulation step.

Reply:
Yes, that is what I meant. I share your concern on the performance impact. I know there is interest in adding more parameters (https://github.com/ignition-forks/dart/pull/22#issuecomment-817376968), but doing so would break ABI unless (a) we add those parameters now, (b) we provide a means for adding them later on without braking ABI

What about a hybrid approach where we add a pointer to extra parameters that is only allocated on demand. For now, the current list of parameters will be directly available in the struct, but new parameters will be added in a new struct hidden to end users and only made accessible via getters/setters.